TBOX-272: Custom UI buttons

Adam-Tamr opened the pull request:
<!---
Thanks for filing a pull request 😄 ! 

Pull requests are only accepted from Tamr employees. If there is a change you would like to see, please file a GitHub Issue or contact your Tamr Professional Services representative.

Below is a checklist of things to do before it can be merged.
-->

# ↪️ Pull Request

<!---
Adding general helper functions for creating yaml files for creating custom Tamr UI buttons
-->


## ✔️ PR Todo

- [x] Add documentation
- [x] Add examples
- [x] Ensure you are following the practices in the [contributor guide](https://docs.google.com/document/d/1c_hhpDMhO0zN793Z_HFOz42Q4g7l1nPRVQ5pXOUsMJU/edit) and [coding_standards](https://datatamr.atlassian.net/wiki/spaces/FEKB/blog/2020/04/14/1137606830/Customer+Success+Coding+Standards)
  * [Use consistent and descriptive name conventions](https://docs.google.com/document/d/1c_hhpDMhO0zN793Z_HFOz42Q4g7l1nPRVQ5pXOUsMJU/edit#heading=h.477s36w4rds9) 
  * [Uses google-style docstrings](https://datatamr.atlassian.net/wiki/spaces/FEKB/blog/2020/04/14/1137606830/Customer+Success+Coding+Standards#Google-style-docstrings)
  * [Uses versioned API/client methods only](https://docs.google.com/document/d/1c_hhpDMhO0zN793Z_HFOz42Q4g7l1nPRVQ5pXOUsMJU/edit#heading=h.yv6df99fyrq2)
  * [Avoid global variables](https://datatamr.atlassian.net/wiki/spaces/FEKB/blog/2020/04/14/1137606830/Customer+Success+Coding+Standards#Avoid-global-variables)
  * [Avoid giant try/catch](https://datatamr.atlassian.net/wiki/spaces/FEKB/blog/2020/04/14/1137606830/Customer+Success+Coding+Standards#Avoid-giant-try-/-except)
  * [Don't use mutable default variables](https://datatamr.atlassian.net/wiki/spaces/FEKB/blog/2020/04/14/1137606830/Customer+Success+Coding+Standards#Don%E2%80%99t-use-mutable-default-arguments)
  * [Use a `main()` function in your scripts](https://datatamr.atlassian.net/wiki/spaces/FEKB/blog/2020/04/14/1137606830/Customer+Success+Coding+Standards#Use-a-%E2%80%9Cmain()%E2%80%9D--function-in-your-scripts)
- [x] Added/updated testing for this change (ensure you are testing any changes/new code!)
- [x] Update title to link to TBOX issue (i.e. start pull request title with TBOX-xx) 
- [x] Approval from first code-reviewer
- [x] Approval from second code-reviewer (only required if a second reviewer has left comments)
- [x] Pass all checks

Adam-Tamr committed:
Sync fork

Adam-Tamr committed:
Revert "Sync fork"

This reverts commit bac5a47148e608767b196c9f0b41bcee2fde86b5.

Adam-Tamr committed:
Test commit

Adam-Tamr committed:
Add custom_buttons.py helper functions for adding custom UI buttons to Tamr

- Create python dicts with UI config attributes
- Save dicts as tamr friendly yaml files
- Group dicts into extension yaml files

Adam-Tamr committed:
Add example button creation script

Adam-Tamr committed:
Format fixes

Adam-Tamr committed:
Format fixes

Adam-Tamr committed:
Add example to docs

Adam-Tamr committed:
Fix doc update formatting

Adam-Tamr committed:
Code refactor

- Functions now save yaml files directly
- Add register_button function which runs unify-admin utils to activate button
- Update example script

Adam-Tamr committed:
Format fixes

Adam-Tamr committed:
Add tests and more validation

- Add pagename validation to create functions
- Add initial unit tests
- Format fixes

Adam-Tamr committed:
Format fix

Adam-Tamr committed:
Fix package imports

Adam-Tamr committed:
Update function names in tests

Adam-Tamr committed:
Refactor code to correct yaml format required

- Add extensionName to every outputted yaml file as it is a requirement
- Add functionality to delete button extension yaml files
- Add test for deletion
- update example script with new args

Adam-Tamr committed:
Return path to yaml files created from button & extension functions

Adam-Tamr committed:
Add function which can manually create extension file from a list of button dicts

Adam-Tamr committed:
Format fix

Adam-Tamr committed:
Add validation to create_extension_from_list func

Adam-Tamr committed:
Merge branch 'Datatamer:develop' into develop

zbpvarun-tamr commented on doc_src/examples/utils.rst:
```diff
@@ -88,4 +88,12 @@ How to get a list of datasets and projects downstream from a specified dataset
+------------------------------------------
+How to create custom UI button yaml files.
+~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+.. literalinclude:: ../../examples/scripts/utils/custom_button/create_custom_buttons.py
```
You should also add a line to doc_src/modules/utils.rst so the documentation from your new file gets added. Also, you should add both in the example and the documentation that this feature is only applicable to versions 2022.008 and up.

Adam-Tamr replied:
Added 👍

zbpvarun-tamr commented on tamr_toolbox/utils/custom_button.py:
```diff
@@ -0,0 +1,422 @@
+    Returns:
+    """
+    LOGGER.info(f"Registering the following buttons in Tamr: {buttons}")
+
```
Question here: In my experience, the folks who use python scripts and such to interface with Tamr do so locally and are separate from those who maintain Tamr infrastructure. In which case, they will have the yaml files created above on their local machines. Will this function work without first transferring the yaml file to the Tamr VM? Per my understanding, it will not...

Adam-Tamr replied:
These functions will only work when ran as scripts **directly on the server**. Which is the most common way I have seen pipelines implemented at customers. In my opinion it would be too complicated to create and get the yaml files onto the servers remotely.

zbpvarun-tamr replied:
I'm debating where the best place to put this would be but we should make it clear somewhere in the docs that these functions need to be run directly on the server to work. Perhaps in the docstring for the register_buttons function? Since the other functions will technically still work even on a local machine. And it would be worth adding a line to that effect at the top of the example script as well.

zbpvarun-tamr commented on tamr_toolbox/utils/custom_button.py:
```diff
@@ -0,0 +1,422 @@
+        )
+
+
+def register_button(
```
If I understand correctly, this function is redundant with the above one with the explicit option of taking a string as input. If so, you can get rid of this function and simply add this to the previous function:
`if isinstance(buttons, str):
    buttons = [buttons]`

Adam-Tamr replied:
Good idea, will change

zbpvarun-tamr commented on tamr_toolbox/utils/custom_button.py:
```diff
@@ -0,0 +1,422 @@
+    """
+    # Check all files exist
+    missing_files = [f for f in button_files if not os.path.exists(f)]
+    if len(missing_files) > 0:
```
Given that people might apply this function iteratively not bothering to delete successfully deleted file paths, I'd say a better option might be to add an info line to the logger that those files were not found and are being skipped instead of raising an error.

Adam-Tamr replied:
Improved `delete_buttons` function to log warning for non present files and continue with deletion of present files.

zbpvarun-tamr commented on tamr_toolbox/utils/custom_button.py:
```diff
@@ -0,0 +1,422 @@
+    for button in buttons:
+        # Run the register command
+        command = (
+            f"{tamr_install_dir}/tamr/utils/unify-admin.sh ui:config --extensionConfig {button}"
```
Does this command run in older versions? Either way, this might be a good place to put in a version check.

Adam-Tamr replied:
Agreed, will add

zbpvarun-tamr commented on examples/scripts/utils/custom_button/create_custom_buttons.py:
```diff
@@ -0,0 +1,82 @@
+        button_id="post_button_1",
+        button_text="Export Project",
+        page_names=["Mastering:Clusters"],
+        post_url="https://example-location.tamr.com/python/api/export",
```
Is this the API to hit for project export?

Adam-Tamr replied:
Just took this from an example doc, will update to a real URL which people would use

Adam-Tamr replied:
See comment below, all examples will be custom python endpoints, so not meaningful for this docs example unfortunately.

zbpvarun-tamr commented on examples/scripts/utils/custom_button/create_custom_buttons.py:
```diff
@@ -0,0 +1,82 @@
+    )
+
+    button_2 = tbox.utils.custom_button.create_redirect_button(
+        extension_name="extension_2",
```
Instead of another random redirect here, perhaps add another post example that users are likely to want? Like a materialize model for categorization on the Schema Mapping page?

Adam-Tamr replied:
Will do 👍

Adam-Tamr replied:
The working examples need to be created via custom python wrappers. See this slack thread https://tamr.slack.com/archives/C0TRASFLL/p1671203404899539?thread_ts=1671202574.750759&cid=C0TRASFLL

zbpvarun-tamr replied:
Oi vey, that is involved! Ok, since I have seen customers use the example scripts as is, I would suggest adding a comment above the POST example that it requires configuring an API wrapper for it to work.

zbpvarun-tamr commented on tests/utils/test_custom_button.py:
```diff
@@ -0,0 +1,57 @@
+)
+import tempfile
+from pathlib import Path
+
```
I see you have tested whether the exceptions you built in are being caught with these tests. But is there a way to test that button creation and function are working as intended? Might need input from other team members on this one...

zbpvarun-tamr replied:
One obvious addition would be to add a yaml file, check that the file exists and is in the format that is expected. But the other stuff may not be possible since it involves a restart of Tamr... Maybe you could model it based on this? https://github.com/Datatamer/tamr-toolbox/blob/develop/tests/sysadmin/test_instance.py#L269

Adam-Tamr replied:
I have personally tested the end to end flow using all the functions on a sandbox and they work as expected. I spoke with Keziah (who made the `test_instance.py` tests) and she explained that they do not actually test that the functions work as expected since they require running the `unify-admin.sh` tool and restarting tamr etc. She explained it would be too much work to properly test this. However I have thoroughly tested on an internal instance manually

Adam-Tamr committed:
Majority of requested changes added

- Added module docs & note that min tamr version is 2022.008
- Removed redundant function
- Add Tamr version check for register_buttons() function
- Added path validation to all functions

Adam-Tamr committed:
Format fixes & Example update

Adam-Tamr committed:
Format fix

Adam-Tamr committed:
Fix tests

Adam-Tamr committed:
Format fix

Adam-Tamr committed:
Fix test

Adam-Tamr committed:
Merge branch 'Datatamer:develop' into develop

Adam-Tamr committed:
Fix testing errors

zbpvarun-tamr reviewed (commented):
I think you're basically there. Would just like you to add a bit more documentation. Mentioned at the relevant previous conversations.

Adam-Tamr committed:
Improve documentation regarding how where to run functions from.

Adam-Tamr committed:
Fix formatting

Adam-Tamr committed:
Fix formatting

zbpvarun-tamr reviewed (dismissed):
Looks good to me!

OGordon100 commented on tamr_toolbox/utils/custom_button.py:
```diff
@@ -0,0 +1,490 @@
+
+    return filepath
+
+
```
It is possible to do automatic version checking at a functional level using [pie decorators](https://github.com/Datatamer/tamr-toolbox/blob/develop/tamr_toolbox/utils/version.py#L149).

So in front of any function you could add `@requires_tamr_version(min_version=2022.008)` in the line before the definition. This looks a bit nicer vs using is_version_condition_met() within the function itself.

Adam-Tamr replied:
Fixed

OGordon100 commented on tamr_toolbox/utils/custom_button.py:
```diff
@@ -0,0 +1,490 @@
+
+    # Page name validation
+    invalid_pages = [p for p in page_names if not _check_valid_page_name(pagename=p)]
+    if len(invalid_pages) > 0:
```
Empty lists are identical to `None`, which evaluates in lists as `False`.

As a result, you can instead just say `if invalid_pages:`

Adam-Tamr replied:
Thanks for the spot, change all occurrences to better syntax

OGordon100 commented on tamr_toolbox/utils/custom_button.py:
```diff
@@ -0,0 +1,490 @@
+        "Schema Mapping:Unified Dataset",
+    ]
+
+    if pagename not in valid_pagenames:
```
Can be simplified down to 1 line:

`return pagename in valid_pagenames`

Adam-Tamr committed:
Minor code improvements & syntax changes

wpmartin reviewed (approved):
lgtm
